src/hawkbit-client.c

```diff
@@ -242,6 +242,68 @@ static gboolean set_auth_curl_header(struct curl_slist **headers, GError **error
         return res;
 }
 
+/**
+ * @brief Set Curl options for TLS/SSL client authentication
+ *
+ * @param[in] curl Curl handle
+ * @param[out] error   Error
+ * @return TRUE if ssl authorization method set in config was set successfully,
+ *      FALSE otherwise (error set)
+ */
+static gboolean set_auth_curl_ssl(CURL *curl, GError **error)
+{
+        curl_easy_setopt(curl, CURLOPT_SSLKEY, hawkbit_config->ssl_key);
+        curl_easy_setopt(curl, CURLOPT_SSLCERT, hawkbit_config->ssl_cert);
+
+        if (!hawkbit_config->ssl_engine)
+                return TRUE;
+
+        if (curl_easy_setopt(curl, CURLOPT_SSLENGINE, hawkbit_config->ssl_engine) != CURLE_OK) {
+                g_set_error(error, RHU_HAWKBIT_CLIENT_CURL_ERROR,
+                            CURLE_FAILED_INIT, "Failed to set ssl engine");
+                return FALSE;
+        }
+        curl_easy_setopt(curl, CURLOPT_SSLKEYTYPE, "ENG");
+        if (curl_easy_setopt(curl, CURLOPT_SSLENGINE_DEFAULT, 1L) != CURLE_OK) {
+                g_set_error(error, RHU_HAWKBIT_CLIENT_CURL_ERROR,
+                            CURLE_FAILED_INIT, "Failed to set engine as default");
+                return FALSE;
+        }
+        g_debug("Using SSL engine %s", hawkbit_config->ssl_engine);
+
+        return TRUE;
+}
+
+/**
+ * @brief Set Curl options for client authentication
+ *
+ * @param[in] curl Curl handle
+ * @param[out] headers curl_slist** of already set headers
+ * @param[out] error   Error
+ * @return TRUE if authorization method set in config and header was added successfully,
+ *      TRUE if no authorization method set, FALSE otherwise (error set)
+ */
+static gboolean set_auth_curl(CURL *curl, struct curl_slist **headers, GError **error)
+{
+        gboolean res;
+
+        // Try ssl authentication
+        if (hawkbit_config->ssl_key && hawkbit_config->ssl_cert) {
+                res = set_auth_curl_ssl(curl, error);
```

Comment:
On failure, this error will be ignored and potentially overridden.

Reply:
@rvdgracht @flobz After thinking about this again, why should we fall back to token authentication if `ssl_key`/`ssl_cert` are given? Shouldn't that be an error? I tend to simply return `res` here.

Reply:
I've dropped the fallback behavior now. I don't think this should block this PR from getting merged.